### DISCUSSION TITLE
Detect if I am connect with a remote device

### DISCUSSION THREAD

**Discussion by triccyx** [robotology/yarp#2580] [Architecture]
With the following code:

```c++
yarp::os::Network yarp;
yarp::dev::PolyDriver device_;
...
yarp::os::Property property;
property.put("device", "remote_grabber");
property.put("local", "/xxx");
property.put("remote", "/grabber/rpc");
if (!device_.open(property))
{
//Error
}

```
I face the following problem if the remote device is not running (yarpdev with grabberDual):
- The open method hangs for some minutes (but not always)
- The open method always exits with true

To me the correct behaviour should be:
- The open method hangs for a selectable timeout (default 1 sec)
- The open method exits with false (in this case)

I'm not able to understand if I am really connected to the remote device.
Am I miss something?

**Comment by drdanz** [comment]
I don't think that there is a generic way to do this...
First of all you can check if YARP network is available:

```c++
yarp::os::Network yarp;
if (!yarp::os::NetworkBase::checkNetwork()) {
    yError() << "yarp server does not seem available";
    return EXIT_FAILURE;
}
```

If you know the name of the remote port(s) that are used by the device (i.e. `remote` in this case), you can also check if they are available:

```c++
if (!yarp::os::NetworkBase::exists("/grabber/rpc")) {
    yError() << "Error: the remote port is not open";
    return EXIT_FAILURE;
}
```

If you are dealing with a synchronization among processes issue, you can also use the `yarp wait /grabbber/rpc` command before your command, that will stop the execution until the port is available.

**Comment by triccyx** [comment]
> If you know the name of the remote port(s) that are used by the device (i.e. remote in this case), you can also check if they are available:

Ok Thanks
